File: hello-grpc-go/common/log_formatter.go
package common

import (
	"context"
	"os"
	"time"

	log "github.com/sirupsen/logrus"
	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/metadata"
	"google.golang.org/grpc/peer"
	"google.golang.org/grpc/status"
)

const serviceName = "go"

var isSecureMode = os.Getenv("GRPC_HELLO_SECURE") == "Y"

// ExtractRequestID extracts request ID from context metadata
func ExtractRequestID(ctx context.Context) string {
	md, ok := metadata.FromIncomingContext(ctx)
	if !ok {
		return "unknown"
	}

	// Try multiple request ID header variants
	if v := md.Get("x-request-id"); len(v) > 0 {
		return v[0]
	}
	if v := md.Get("request-id"); len(v) > 0 {
		return v[0]
	}
	return "unknown"
}

// ExtractPeer extracts peer address from context
func ExtractPeer(ctx context.Context) string {
	if p, ok := peer.FromContext(ctx); ok {
		return p.Addr.String()
	}
	return "unknown"
}

// IsSecure checks if the connection is secure
func IsSecure(ctx context.Context) bool {
	if p, ok := peer.FromContext(ctx); ok && p.AuthInfo != nil {
		return true
	}
	return isSecureMode
}

// LogRequestStart logs the start of an RPC request
func LogRequestStart(ctx context.Context, method string) {
	requestID := ExtractRequestID(ctx)
	peerAddr := ExtractPeer(ctx)
	secure := IsSecure(ctx)

	log.WithFields(log.Fields{
		"service":    serviceName,
		"request_id": requestID,
		"method":     method,
		"peer":       peerAddr,
		"secure":     secure,
		"status":     "STARTED",
	}).Info("rpc_request")
}

// LogRequestEnd logs the completion of an RPC request
func LogRequestEnd(ctx context.Context, method string, startTime time.Time, err error) {
	requestID := ExtractRequestID(ctx)
	peerAddr := ExtractPeer(ctx)
	secure := IsSecure(ctx)
	durationMs := time.Since(startTime).Milliseconds()

	fields := log.Fields{
		"service":     serviceName,
		"request_id":  requestID,
		"method":      method,
		"peer":        peerAddr,
		"secure":      secure,
		"duration_ms": durationMs,
	}

	if err != nil {
		st, _ := status.FromError(err)
		fields["status"] = st.Code().String()
		fields["error_code"] = st.Code().String()
		fields["message"] = st.Message()
		log.WithFields(fields).Error("rpc_response")
	} else {
		fields["status"] = codes.OK.String()
		log.WithFields(fields).Info("rpc_response")
	}
}

// StreamLoggingInterceptor returns a gRPC interceptor that logs streaming RPC calls with unified format
func StreamLoggingInterceptor() grpc.StreamServerInterceptor {
	return func(srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
		startTime := time.Now()
		ctx := ss.Context()

		LogRequestStart(ctx, info.FullMethod)

		err := handler(srv, ss)

		LogRequestEnd(ctx, info.FullMethod, startTime, err)

		return err
	}
}

File: hello-grpc-rust/src/landing/log_formatter.rs
use log::{info, error};
use std::time::Instant;
use tonic::{Request, metadata::MetadataMap, Status};

const SERVICE_NAME: &str = "rust";

/// Extract request ID from metadata
pub fn extract_request_id(metadata: &MetadataMap) -> String {
    // Try multiple request ID header variants
    if let Some(value) = metadata.get("x-request-id") {
        if let Ok(s) = value.to_str() {
            return s.to_string();
        }
    }
    if let Some(value) = metadata.get("request-id") {
        if let Ok(s) = value.to_str() {
            return s.to_string();
        }
    }
    "unknown".to_string()
}

/// Extract peer address from request
pub fn extract_peer<T>(request: &Request<T>) -> String {
    request
        .remote_addr()
        .map(|addr| addr.to_string())
        .unwrap_or_else(|| "unknown".to_string())
}

/// Check if connection is secure
pub fn is_secure() -> bool {
    std::env::var("GRPC_HELLO_SECURE")
        .map(|v| v == "Y")
        .unwrap_or(false)
}

/// Log request start
pub fn log_request_start(method: &str, request_id: &str, peer: &str) {
    let secure = is_secure();
    info!(
        "service={} request_id={} method={} peer={} secure={} status=STARTED",
        SERVICE_NAME, request_id, method, peer, secure
    );
}

/// Log request end
pub fn log_request_end(
    method: &str,
    request_id: &str,
    peer: &str,
    start_time: Instant,
    status: &str,
) {
    let secure = is_secure();
    let duration_ms = start_time.elapsed().as_millis();
    info!(
        "service={} request_id={} method={} peer={} secure={} duration_ms={} status={}",
        SERVICE_NAME, request_id, method, peer, secure, duration_ms, status
    );
}

/// Log request error
pub fn log_request_error(
    method: &str,
    request_id: &str,
    peer: &str,
    start_time: Instant,
    status: &Status,
) {
    let secure = is_secure();
    let duration_ms = start_time.elapsed().as_millis();
    let error_code = format!("{:?}", status.code());
    let message = status.message();
    
    error!(
        "service={} request_id={} method={} peer={} secure={} duration_ms={} status={} error_code={} message={}",
        SERVICE_NAME, request_id, method, peer, secure, duration_ms, error_code, error_code, message
    );
}
